common/quotas/global/algorithm/requestweighted_test.go

```diff
@@ -0,0 +1,746 @@
+		require.Len(t, agg.usage, 0) // also frees memory
+	})
+	t.Run("retains recent data while cleaning", func(t *testing.T) {
+		agg, tick := setup(t)
+
+		// refresh data for one host
+		agg.Update(h1, map[Limit]Requests{key: {1, 1}}, time.Second)
+		tick.Advance(time.Second) // advance to 10th second
+
+		// read both hosts.  h1 should exist, h2 should not
+		weights, used := agg.HostWeights(h1, []Limit{key})
+		require.Len(t, weights, 1, "h1 was refreshed and should remain")
+		require.Len(t, used, 1, "h1 was refreshed and usage data should remain")
+		weights, used = agg.HostWeights(h2, []Limit{key})
+		require.Len(t, weights, 0, "h2 should have no weight at all")
+		require.Len(t, used, 1, "h1 was refreshed and usage data should remain")
+
+		// internals should also be partly emptied
+		require.Len(t, agg.usage, 1, "limit should remain")
+		require.Len(t, agg.usage[key], 1, "only one host's data should be in limit")
+	})
+	t.Run("cleans up by explicit gc", func(t *testing.T) {
+		agg, tick := setup(t)
+		tick.Advance(time.Second)
+		met := agg.GC()
+		require.Equal(t, Metrics{
+			HostLimits: 0, // none remain
+			Limits:     0, // none remain
+
+			RemovedHostLimits: 2, // h1 and h2 for the single key, both removed
+			RemovedLimits:     1, // single key removed
+		}, met)
+
+		// internals should also be empty
+		require.Len(t, agg.usage, 0)
+	})
+}
+
+func TestRapidlyCoalesces(t *testing.T) {
+	// This test ensures that, regardless of starting weights, the algorithm
+	// "rapidly" achieves near-actual weight distribution after a small number of rounds.
+	//
+	// the exact numbers here don't really matter, it's just handy to show the behavior
+	// in semi-extreme scenarios.  logs show a quick adjustment which is what we want.
+	// if making changes, check with like 10k rounds to make sure it's stable.
+	//
+	// time is also not advanced because it doesn't actually need to be advanced.
+	// an update is an update, and the caller's elapsed time is assumed to be correct.
+	agg, _ := newForTest(t, defaultConfig(time.Second))
+
+	key := Limit("start workflow")
+	h1, h2, h3 := Identity("one"), Identity("two"), Identity("three")
+
+	weights, used := agg.getWeightsLocked(key, agg.snapshot())
+	assert.Zero(t, weights, "should have no weights")
+	assert.Zero(t, used, "should have no used RPS")
+
+	push := func(host Identity, accept, reject int) {
+		agg.Update(host, map[Limit]Requests{
+			key: {
+				Accepted: accept,
+				Rejected: reject,
+			},
+		}, time.Second) // 1s just to make rps in == rps out
+	}
+
+	// init with anything <~1000, too large and even a small fraction of the original value can be too big.
+	push(h1, rand.Intn(1000), rand.Intn(1000))
+	push(h2, rand.Intn(1000), rand.Intn(1000))
+	push(h3, rand.Intn(1000), rand.Intn(1000))
+
+	// now update multiple times and make sure it gets to 90% within 4 steps == 12s (normally).
+	//
+	// 4 steps with 0.5 weight should mean only 0.5^4 => 6.25% of the original influence remains,
```

Comment:
nit: picking 0.5 weight makes `1 - weight == weight` and can be misread here. for example 0.5^4 in the comment is actually referring to (1-0.5)^4. 
It makes more sense to give higher weight to new data as we don't want to reject calls as a punishment to high past rps count too much. 
So we can go with default weight=0.7 to address these.
Actual sane number will be determined after real world tests of course.

Reply:
yep, I was thinking this last night too - definitely agreed, I should use non-0.5 pretty much universally in tests.